server/server.go

```diff
@@ -808,12 +855,53 @@ type handlerSwitcher struct {
 	contentTypeToHandler map[string]http.Handler
 }
 
+// Add JSON API paths to NoCsrfHeaderPatterns slice that allow unauthenticated
+// GET requests, so that they will not return X-CSRF-Token header on response.
+var NoCsrfHeaderPattern = []*regexp.Regexp{
+	regexp.MustCompile(`/api/version.*`),
+	regexp.MustCompile(`/api/v1/settings.*`),
+}
+
 func (s *handlerSwitcher) ServeHTTP(w http.ResponseWriter, r *http.Request) {
 	if urlHandler, ok := s.urlToHandler[r.URL.Path]; ok {
 		urlHandler.ServeHTTP(w, r)
 	} else if contentHandler, ok := s.contentTypeToHandler[r.Header.Get("content-type")]; ok {
 		contentHandler.ServeHTTP(w, r)
 	} else {
+		// If we have a valid authorization header for API access, skip CSRF protection for this request
+		// This is accomplished by setting the key gorilla.csrf.Skip in the request's context.
+		//
+		// We don't validate the JWT token here - this is done at API level. But we make sure that the
+		// request does not carry a cookie named "argocd.token" as sent by the browser. This would
+		// circumvent CSRF protection.
+		//
+		// API access from scripts or other clients therefore MUST always set "Authorization" header
+		// and MUST NOT send the JWT in a cookie (or MUST ALSO send the X-CSRF-Token header along)
+		//
+		// We do allow a POST on /api/v1/session to get a valid JWT token without having to also
+		// supply the X-CSRF-Token header, as long as the argocd.token is not set.
+		//
+		if _, err := r.Cookie(common.AuthCookieName); err != nil {
+			if authHdr := r.Header.Get("Authorization"); strings.HasPrefix(authHdr, "Bearer ") {
+				r = csrf.UnsafeSkipCheck(r)
```

Comment:
What happens when a request has both a cookie, and a auth header that looks like this:

    Authorization: Bearer x 

In that case, the CSRF protection will be disabled because of this code, and getToken will use any valid token, either in the cookie or in the header. It seems that this way it is possible to disable CSRF protection while keeping the token in the cookie.

Reply:
My assumptions on this are as follows:

If both, cookie and authorization header, carry an invalid token, the user will not be authenticated. Unauthenticated, they will not be able to perform any of the data manipulation methods in the underlying gRPC services, except when authentication on the server is disabled, but then we don't need CSRF checks anyway IMHO. During the time the request goes through the HTTP-to-gRPC gateway proxy, the token - whether passed through cookie or Authorization header - was not yet validated anywhere. This validation happens at the gRPC layer.

The check for `_, err := r.Cookie(...); err != nil` will evaluate to true if the request does not carry a cookie with name `argocd.token` and the gRPC layer will ultimately only be able to validate the token in the `Authorization` header. If both, Cookie and Token carry a valid token, CSRF protection should kick in.

If I assumed wrongly, I'm happy to discuss further :)

Reply:
It seems I misunderstood the cookie check. So this will only disable CSRF when there is no cookie, which is indeed the correct behavior.